logstore: sync sideloaded storage directories

pav-kv opened the pull request:
This PR ensures that the hierarchy of directories/files created by the sideloaded log storage is properly synced.

Previously, only the "leaf" files in this hierarchy were fsync-ed. Even though this guarantees the files content and metadata is synced, this still does not guarantee that the references to these files are durable. For example, Linux man page for `fsync` [^1] says:

```
Calling fsync() does not necessarily ensure that the entry in the
directory containing the file has also reached disk.  For that an
explicit fsync() on a file descriptor for the directory is also
needed.
```

It means that these files can be lost after a system crash of power off. This leads to issues because:

1. Pebble WAL syncs are not atomic with the sideloaded files syncs. It is thus possible that raft log metadata "references" a sideloaded file and gets synced, but the file is not yet. A power off/on at this point leads to an internal inconsistency, and can result in a crash loop when raft will try to load these entries to apply and/or send to other replicas.

2. The durability of entry files is used as a pre-condition to sending raft messages that trigger committing these entries. A coordinated power off/on on a majority of replicas can thus lead to losing committed entries and unrecoverable loss-of-quorum.

This PR fixes the above issues, by syncing parents of all the directories and files that the sideloaded storage creates.

[^1]: https://man7.org/linux/man-pages/man2/fsync.2.html

Part of #114411

Epic: none

Release note (bug fix): this commit fixes a durability bug in raft log storage, caused by incorrect syncing of filesystem metadata. It was possible to lose writes of a particular kind (`AddSSTable`) that is e.g. used by `RESTORE`. This loss was possible only under power-off or OS crash conditions. As a result, CRDB could enter a crash loop on restart. In the worst case of a correlated power-off/crash across multiple nodes this could lead to loss of quorum or data loss.

pav-kv committed:
logstore: sync sideloaded directories creation

The sideloaded log storage does not sync the hierarchy of directories it
creates. This can potentially lead to full or partial loss of its
sub-directories in case of a system crash or power off.

After this commit, every time sideloaded storage creates a directory, it
syncs its parent so that the reference is durable.

Epic: none
Release note: none

pav-kv committed:
logstore: rm test-only DiskSideloadStorage field

This commit removes the dirCreated field of DiskSideloadStorage, because
it is only used in tests, and is reduntant (directory existence check
already does the job).

Epic: none
Release note: none

pav-kv committed:
logstore: inline dir creation in DiskSideloadStorage

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114616)
<!-- Reviewable:end -->

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload.go:
```diff
@@ -140,8 +143,12 @@ func MaybeSideloadEntries(
-		// We never saw a sideloaded command.
+	if output != nil { // there is at least one sideloaded command
+		// Sync the sideloaded storage directory so that the commands are durable.
+		if err := sideloaded.Sync(); err != nil {
```
This sync could be moved up the stack, to `LogStore.storeEntriesAndCommitBatch` or higher if we want a delayed/"async" `Sync`. The sideloaded storage syncs can't be batched across ranges/replicas though (because each one is on a separate directory), so there is not much sense in moving it up right now.

pavelkalinnikov replied:
Each `Put()` syncs the added file, but not the containing directory.

`Sync()` syncs the directory containing these files (so the references to the newly added files are persisted), and is called once per `storeEntriesAndCommitBatch` (which is called once per raft ready run). I.e. it "batches" things to some extent, if there are multiple `Put()` entries in a batch.

Other directories (parents of the per-range directories; i.e. the entire path like `auxiliary/sideloading/r2XXX`) are synced once in `mkdirAllAndSyncParents` when they are created. After that, they don't need to be synced again.

erikgrinaker replied:
I don't understand. Consider this sequence, using `auxiliary/sideloaded` as the base directory:

```
Put("r1XXX/r1234")
Sync()
```

`Sync()` will only sync `auxiliary/sideloaded`, so it won't guarantee durability of `r1XXX/r1234`'s entry in `r1XXX`, only `r1XXX`'s entry in the base dir.

If `Put()` is supposed to guarantee durability of `r1XXX/r1234`'s directory entry, then it should presumably also guarantee durability of `somefile` in the base dir, in which case `Sync()` is redundant.

What am I missing?

erikgrinaker replied:
Nevermind, we instantiate the sideloaded storage per replica, so it's always a flat directory.

pavelkalinnikov replied:
The base directory is `auxiliary`, and the full path to the sideloaded storage for range 1234 is `auxiliary/sideloaded/r1XXX/r1234`.

When a sideloaded storage for range 1234 is created, it needs to create and sync the entire `auxiliary/sideloaded/r1XXX/r1234` path. But we don't do it instantly on creation, instead it's done lazily when `Put` writes a first entry file. If `Put` discovers that the directories are missing, it creates and syncs them. But we can pretty much assume that it happens only once.

The normal operation of `Put` only adds files to an already existing `auxiliary/sideloaded/r1XXX/r1234` directory. It does so by calling [WriteFileSyncing](https://github.com/cockroachdb/cockroach/blob/dcedaf1cc617796bce6f8313f74631e1c0dbcad8/pkg/kv/kvserver/logstore/sideload_disk.go#L100) which fsyncs the file's content.

After we've run a bunch of `Put`s, everything is synced except the `.../r1234` folder (aka the [dir](https://github.com/cockroachdb/cockroach/blob/dcedaf1cc617796bce6f8313f74631e1c0dbcad8/pkg/kv/kvserver/logstore/sideload_disk.go#L42C1-L42C1) field), because it now contains some new files. That's why we need the additional `Sync()` method, so sync this `ss.dir` folder.

pavelkalinnikov replied:
I split out a couple more commits to demonstrate the effect of multiple syncs in this PR. See the evolution of `TestSideloadStorageSync` across the commits.

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload_disk.go:
```diff
@@ -276,3 +293,57 @@ func (ss *DiskSideloadStorage) String() string {
+//
+// NB: path syntax in the examples above may differ, because components of the
+// path are combined using the provided filesystem interface.
+func mkdirAllAndSyncParents(fs vfs.FS, base string, dirs []string, perm os.FileMode) error {
```
This is a modified version of a similar [func](https://github.com/cockroachdb/pebble/blob/124486b0f85d75a0c695cb65703c5495eec66c99/checkpoint.go#L87) in pebble.

I trust this implementation more though, because:
- there is no sketchy "." checks (which, if done wrong, can lead to infinite loops)
- there is a unit-test

@cockroachdb/storage consider upstreaming this and/or the tests.

erikgrinaker replied:
Any compelling reason why this doesn't simply take the path as a string, without special-casing `base`, and traverse it using e.g. `os.filepath.Dir()` until it finds an existing directory?

I see you mention sketchy `.` checks, but what's preventing the caller from passing in a path with `.` here anyway? Seems better to just handle that case.

pavelkalinnikov replied:
The other implementation does:
```
while path != "." {
	process(path)
	path = parent(path) // who guarantees that this will always return "." in the end?
}
```

The current implementation always terminates because the path is already decomposed and has a predefined number of elements.

pavelkalinnikov commented:
Recommend reviewing commit-by-commit.

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload_disk.go:
```diff
@@ -97,13 +101,26 @@ func (ss *DiskSideloadStorage) Put(
 
+// Sync implements SideloadStorage.
+func (ss *DiskSideloadStorage) Sync() error {
+	dir, err := ss.eng.OpenDir(ss.dir)
+	if err != nil {
+		return err
+	}
```
There are some cases when `ss.dir` is not yet, or already not existing. This code must account for that.

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload_disk.go:
```diff
@@ -285,3 +293,57 @@ func (ss *DiskSideloadStorage) String() string {
+// For example, if dirs is [x, y, z], then this func creates all the missing
+// subdirectories in base/x/y/z. If base is the lowest previously existing
+// directory then base, x, and y are synced. If x existed too, then only x and y
+// are synced. If y existed before then there are no syncs.
```
typo: If y existed -> If z existed

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload_disk.go:
```diff
@@ -102,15 +99,28 @@ func (ss *DiskSideloadStorage) Put(
-		if err := ss.createDir(); err != nil {
+		// Ensure that ss.dir exists. The filename() is placed directly in ss.dir,
+		// so the next loop iteration should succeed.
+		if err := mkdirAllAndSyncParents(ss.eng, ss.path[0], ss.path[1:], os.ModePerm); err != nil {
```
NB: this will sync parents only for the newly created directories. For instance:

- if `auxiliary/sideloading/r1XXX/r1234` already exists then there will be no sync
- if `auxiliary/sideloading/r2XXX` exists and we add `auxiliary/sideloading/r2XXX/r2222`, then only the `r2XXX` folder is synced.

This leaves a window of vulnerability in existing clusters who already have sideloaded files and did not sync the folders before. Only **newly** created sideloaded storages are strongly synced.

We could make this stronger for the existing clusters too, by ensuring that every pre-existing folder is synced at least once. Some ways of doing this:

1. add a migration that traverses and syncs the sideloaded storage at startup
2. or, lazily force a sync once per directory (but we would need the overhead of tracking which ones are [not] synced)
3. or, call something like [syncfs](https://linux.die.net/man/2/syncfs) on startup.

However, in practice, only recently created directories are susceptible to "disappearing" after a power outage. I think it's fine to accept this little risk, and not introduce any migration complexity. The sideloaded storage directories will be fully synced when a binary with this patch is started, **and** a bit of time has passed so that the FS has flushed all the pre-existing dirs. After that, the storage is always synced. I.e. time is our "migration".

erikgrinaker replied:
I don't think we need to consider existing clusters here. Since the Raft state and log are stored in the same Pebble instance, I believe applied state changes (i.e. the SST hardlink) will be synced on the next Raft append, which will happen shortly after a restart with the new binary. Does that sound right?

pavelkalinnikov replied:
Yeah, the SSTs are hard linked into Pebble on apply stage when doing the ingestion (which also syncs the necessary parts of the FS): https://github.com/cockroachdb/cockroach/blob/14733120c08ba1a0a050bb5442c3311090206566/pkg/kv/kvserver/replica_proposal.go#L724

But, I think, the references to these files from the log's directory may still remain un-synced. So if we restart after a crash, it's possible that the entries are "gone" from the log. Raft may try to access these log entries (e.g. to apply them again, or to catch up other followers).

pavelkalinnikov replied:
Also, this argument only applies to SSTs that were applied. The unapplied ones are more vulnerable.

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload_disk.go:
```diff
@@ -102,15 +99,28 @@ func (ss *DiskSideloadStorage) Put(
 }
 
+// Sync implements SideloadStorage.
+func (ss *DiskSideloadStorage) Sync() error {
```
NB: this method implementation does two syscalls: `OpenDir` and `Sync`.

I considered calling `OpenDir` once, around the time when the directories are opened/created, and storing the handle in the struct. Then `Sync` could reusing this handle, and do just a single syscall.

However, the handle holds a file descriptor. There is one sideloaded storage instance per `Range`, so correspondingly we may have many descriptors open at once. Since the descriptors are a limited resource, I decided against this approach.

erikgrinaker reviewed (changes requested):
Have we run any benchmarks to look at the performance impact?

pavelkalinnikov commented on pkg/kv/kvserver/replica_destroy.go:
```diff
@@ -84,6 +84,10 @@ func (r *Replica) postDestroyRaftMuLocked(ctx context.Context, ms enginepb.MVCCS
 	// the replica is always contained in its descriptor. So this code below should
 	// be removable.
+	//
+	// TODO(pavelkalinnikov): coming back in 2023, the above may still happen if:
+	// (1) state machine syncs, (2) OS crashes before (3) sideloaded was able to
+	// sync the files removal. The files should be cleaned up on restart.
```
Yeah, should be a quick win. Assigned #26813 to myself.

pav-kv committed:
logstore: add TODOs in sideloaded storage

A couple of things to address in the future: sideloaded files removal
should happen strictly after a state machine sync; sideloaded files and
directories should be cleaned up on startup because their removal is not
always durable.

Epic: none
Release note: none

pavelkalinnikov commented:
> Have we run any benchmarks to look at the performance impact?

No. Some of `AddSST`-heavy roachtest would do, like `restore` ones?

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload_test.go:
```diff
@@ -571,3 +573,173 @@ func newOnDiskEngine(ctx context.Context, t *testing.T) (func(), storage.Engine)
+		defer eng.Close()
+		ss = newTestingSideloadStorage(eng)
+
+		// The stored entry is still durable if we synced the sideloaded storage
+		// before the crash.
+		got, err := ss.Get(ctx, 100 /* index */, 6 /* term */)
+		if sync {
+			require.NoError(t, err)
+			require.Equal(t, testEntry, string(got))
```
Before this change, this test would fail, i.e. we would lose the entry.

erikgrinaker commented:
> > Have we run any benchmarks to look at the performance impact?
> 
> No. Some of `AddSST`-heavy roachtest would do, like `restore` ones?

Yeah, merging and looking at roachperf nightlies is probably sufficient. But i would be good to have a microbenchmark too, if we don't already have one, to look at the worst-case impact.

> We already [sync](https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvserver/kvserverbase/syncing_write.go#L119) every [512 KiB](https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvserver/kvserverbase/syncing_write.go#L66-L72) when writing each sideloaded file. After this PR we will sync the parent directory of this file too, so that's K+1 syncs instead of K. This might have some impact, but not much if the `AddSST`s are multi-MiB in size. If this becomes a problem, one can double `kv.bulk_sst.sync_size` to 1MiB.

Yeah, the directory fsyncs are also tiny and won't block on any other writes (unlike e.g. the WAL fsyncs).

> This change doesn't feel optional though. It's a correctness/durability bug. We could optimize it later by having a better log storage (e.g. batch multiple entries into a single file since this is mostly append-only; but then would Pebble be able to ingest partial files with the current API?).

Yeah, I agree that this isn't optional, but knowing the perf impact would give us some idea of what to expect from a backport. If the impact is severe (I don't expect it to be) then we may be more hesitant to backport this to a patch release.

We could also consider simply spawning a separate goroutine to do the directory fsync concurrently with the SST fsync, which should ~eliminate the latency cost. Only if we need to though.

pavelkalinnikov commented:
> Yeah, merging and looking at roachperf nightlies is probably sufficient. But i would be good to have a microbenchmark too, if we don't already have one, to look at the worst-case impact.

I'm inclined to think we should be fine with the existing benchmarks.

Not sure microbenchmarking is the best framework for testing this kind of change. What kind of microbenchmarks do you mean? Typically microbenchmarks I think of are in-memory, and geared towards measuring things like CPU and memory usage. In this case it wouldn't give us much: the cost of this PR is one extra `OpenDir` + `Sync` (in cases when we commit `AddSST` entries) which is negligible with the in-memory `FS`, the real impact is in IO.

Maybe we could do a benchmark over a real on-disk storage, but then to make it "worst case" we need to put quite an effort into this test. We could do a simple thing like a single-range write-many-AddSSTs, but it wouldn't be stressing the system enough. Instead, we could just piggyback on the `restore` tests which are closer to the real worst case?

Here are a few metrics before/after this change, on the `restore/tpce/400GB/gce/nodes=4/cpus=8` test. The differences are within an error margin.

<details>
<summary>Storage WAL fsync 12ms </summary>

<img width="1065" alt="Screenshot 2023-11-29 at 20 53 36" src="https://github.com/cockroachdb/cockroach/assets/3757441/75d02aee-2acb-4591-8a02-3f23c8e1a506">
<img width="1064" alt="Screenshot 2023-11-29 at 20 54 06" src="https://github.com/cockroachdb/cockroach/assets/3757441/14a7f68a-d205-4208-8403-f098dc73f822">

</details>

<details>
<summary>Write bytes 120 M/s</summary>

<img width="1065" alt="Screenshot 2023-11-29 at 20 54 22" src="https://github.com/cockroachdb/cockroach/assets/3757441/241715ad-b1d0-48bb-bec0-a36437b4c45d">
<img width="1066" alt="Screenshot 2023-11-29 at 20 54 30" src="https://github.com/cockroachdb/cockroach/assets/3757441/aadae570-2339-4086-a565-fde8841e50d1">

</details>

<details>
<summary> Raft batch apply latency 18ms </summary>

https://github.com/cockroachdb/cockroach/blob/933c4cd42da83679fcd460fcdd68ecaee0ad7b10/pkg/kv/kvserver/metrics.go#L1305

<img width="1070" alt="Screenshot 2023-11-29 at 20 54 41" src="https://github.com/cockroachdb/cockroach/assets/3757441/6e3a116e-80da-4d2a-9a92-998fd1794577">
<img width="1065" alt="Screenshot 2023-11-29 at 20 54 54" src="https://github.com/cockroachdb/cockroach/assets/3757441/af637d84-d25a-48ba-86a8-26caff50eb4a">

</details>

The "after" even won the benchmark, although I think this is just [not a particularly stable](https://roachperf.crdb.dev/?filter=restore&view=restore%2Ftpce%2F400GB%2Fgce%2Fnodes%3D4%2Fcpus%3D8&tab=gce) metric:

```
$ grep Throughput: *txt
after-2.txt:; Throughput: 88.786256 mb / node / second
after.txt:; Throughput: 83.704015 mb / node / second
before-2.txt:; Throughput: 79.064119 mb / node / second
before.txt:; Throughput: 62.075737 mb / node / second
```

pav-kv committed:
logstore: sync sideloaded files creation

The sideloaded storage fsyncs the files that it creates. Even though
this guarantees durability of the files content and metadata, this still
does not guarantee that the references to these files are durable. For
example, Linux man page for fsync [^1] says:

```
Calling fsync() does not necessarily ensure that the entry in the
directory containing the file has also reached disk.  For that an
explicit fsync() on a file descriptor for the directory is also
needed.
```

It means that these files can be lost after a system crash of power off.
This leads to issues:

1. The storage syncs are not atomic with the sideloaded files syncs. It
   is thus possible that raft log metadata "references" a sideloaded
   file and gets synced, but the file is not yet. A power off/on at
   this point leads to an internal inconsistency, and can result in a
   crash loop when raft will try to load these entries to apply and/or
   send to other replicas.

2. The durability of entry files is used as a pre-condition to sending
   raft messages that trigger committing these entries. A coordinated
   power off/on on a majority of replicas can thus lead to losing
   committed entries and unrecoverable loss-of-quorum.

This commit fixes the above issues, by syncing the parent directory
after writing sideloaded entry files. The natural point for this is
MaybeSideloadEntries on the handleRaftReady path.

[^1]: https://man7.org/linux/man-pages/man2/fsync.2.html

Epic: none

Release note (bug fix): this commit fixes a durability bug in raft log
storage, caused by incorrect syncing of filesystem metadata. It was
possible to lose writes of a particular kind (AddSSTable) that is e.g.
used by RESTORE. This loss was possible only under power-off or OS crash
conditions. As a result, CRDB could enter a crash loop on restart. In
the worst case of a coordinated power-off/crash across multiple nodes
this could lead to an unrecoverable loss of quorum.

pav-kv committed:
logstore: repro files loss in sideloaded storage

This commit adds `TestSideloadStorageSync` which demonstrates that the
sideloaded log storage can lose files and directories upon system crash.
This is due to the fact that the whole directory hierarchy is not
properly synced when the directories and files are created.

A typical sideloaded storage file (entry 123 at term 44 for range r1234)
looks like: `<data-dir>/auxiliary/sideloading/r1XXX/r1234/i123.t44`.

Only existence of auxiliary directory is persisted upon its creation, by
syncing the <data-dir> when Pebble initializes the store. All other
directories (sideloading, r1XXX, r1234) are not persisted upon creation.

Epic: none
Release note: none

pavelkalinnikov commented on pkg/kv/kvserver/logstore/sideload_disk.go:
```diff
@@ -279,3 +283,57 @@ func (ss *DiskSideloadStorage) String() string {
+			return errors.Wrapf(err, "could not get dir info: %s", paths[i])
+		}
+	}
+	if existing == -1 {
+		return errors.Newf("base dir does not exist: %s", base)
+	}
```
This can be problematic if someone drops the `auxiliary` directory for fun (for example, they see it's empty and assume it is then safe to remove it).

Maybe we should walk all the way up to the Pebble root directory instead, to be more robust.

nvb reviewed (commented):
> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 1 of 6 files at r3, 5 of 5 files at r4, 4 of 4 files at r6, 3 of 3 files at r7, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cockroachdb, @erikgrinaker, @pavelkalinnikov, and @RaduBerinde)

___
*[`-- commits` line 50 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114616#-NkWiB6S48fwYzjE0IK4):*
partial
___
*[`pkg/kv/kvserver/logstore/sideload.go` line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/114616#-NkWnwAQ9sm7uccez49O:-NkWnwAQ9sm7uccez49P:b-kqwzli) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4dce4e66328d309bc484840d3c00c8294aa31a0/pkg/kv/kvserver/logstore/sideload.go#L37)):*
> ```Go
> 	// term. Overwrites the file if it already exists.
> 	Put(_ context.Context, index kvpb.RaftIndex, term kvpb.RaftTerm, contents []byte) error
> 	// Sync syncs the underlying filesystem metadata so that all the preceding
> ```

Why expose `Sync` on the `SideloadStorage` interface instead of pushing the syncs into the `Put`/`Clear`/etc. methods, like we've done for the directory creation syncs?
___
*[`pkg/kv/kvserver/logstore/sideload_disk.go` line 300 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114616#-NkWl_Tv6OYDxRH_5yz_:-NkWl_Tv6OYDxRH_5yza:b8jpvxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/06a82d14da8ff900b392efc43e3fe893bbeb8b76/pkg/kv/kvserver/logstore/sideload_disk.go#L300)):*
> ```Go
> // path are combined using the provided filesystem interface.
> func mkdirAllAndSyncParents(fs vfs.FS, base string, dirs []string, perm os.FileMode) error {
> 	paths := append(make([]string, 0, len(dirs)+1), base)
> ```

We don't need to complicate the code if it doesn't feel worthwhile to you, but enumerating all sub-paths and constructing individual path strings to test feels a bit wasteful. I wonder whether iterative calls to [`path.Split`](https://pkg.go.dev/path#Split) over a precomputed `path` string would make this logic easier to reason about.
___
*[`pkg/kv/kvserver/logstore/sideload_disk.go` line 317 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114616#gh-1409999778:-NkWjwqv0DypULfMb7CM:b1mc2ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/06a82d14da8ff900b392efc43e3fe893bbeb8b76/pkg/kv/kvserver/logstore/sideload_disk.go#L317)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

This can be problematic if someone drops the `auxiliary` directory for fun (for example, they see it's empty and assume it is then safe to remove it).

Maybe we should walk all the way up to the Pebble root directory instead, to be more robust.
</blockquote></details>

Walking all the way up to the Pebble root directory and not assuming that "auxiliary" exists makes sense to me.

Or alternatively we could remove the distinction between the base directory path and the sub-directories we are trying to create and just create+sync whatever is necessary without making assumptions. This is analogous to what MkdirAll does.
___
*[`pkg/kv/kvserver/logstore/sideload_test.go` line 594 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114616#-NkWi2gF-lFjAUdcb8rh:-NkWi2gG0ZkvhiBvh7jy:boz54fp) ([raw file](https://github.com/cockroachdb/cockroach/blob/72995ed527b32d819a672b79d73116dffa3124a3/pkg/kv/kvserver/logstore/sideload_test.go#L594)):*
> ```Go
> 	const testEntry = "test-entry"
> 	require.NoError(t, ss.Put(ctx, 100 /* index */, 6 /* term */, []byte(testEntry)))
> 	// Cut off all syncs from this point, to emulate a crash.
> ```

Why is this needed? Does `eng.Close()` try to sync before closing?


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
>Could you please do another pass when you have a multi-minute?

Won't have time today unfortunately, will have a look next week.

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @nvanbenschoten, and @RaduBerinde)

___
*[`pkg/kv/kvserver/logstore/sideload.go` line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/114616#-NkWnwAQ9sm7uccez49O:-NkoWvXa5cGP1lpdESKt:brl2y47) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4dce4e66328d309bc484840d3c00c8294aa31a0/pkg/kv/kvserver/logstore/sideload.go#L37)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why expose `Sync` on the `SideloadStorage` interface instead of pushing the syncs into the `Put`/`Clear`/etc. methods, like we've done for the directory creation syncs?
</blockquote></details>

This is to enable some batching. We can `Put` multiple entries, and then `Sync` the directory once to persist the addition of all the entries at once.
___
*[`pkg/kv/kvserver/logstore/sideload_disk.go` line 309 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114616#gh-1396484019:-NkotnWtDOqBBpEgNo5l:bjyeptn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4688eb20b690226266d74acd7692857c5744f2b/pkg/kv/kvserver/logstore/sideload_disk.go#L309)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

The other implementation does:
```
while path != "." {
	process(path)
	path = parent(path) // who guarantees that this will always return "." in the end?
	// also, all vfs.FS implementations must return this "." which is a lot to expect

	// on top of that, I ran some tests, and parent(path) sometimes did
	// not return ".", and had special effects depending on whether the path
	// is absolute or relative
}
```

The current implementation always terminates because the path is already decomposed and has a predefined number of elements.
</blockquote></details>

Changed the code to walk directories up using the `fs.PathDir()` method.
___
*[`pkg/kv/kvserver/logstore/sideload_disk.go` line 300 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114616#-NkWl_Tv6OYDxRH_5yz_:-Nkou-PW3EnM97Eus2lZ:b-vrjrl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/06a82d14da8ff900b392efc43e3fe893bbeb8b76/pkg/kv/kvserver/logstore/sideload_disk.go#L300)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We don't need to complicate the code if it doesn't feel worthwhile to you, but enumerating all sub-paths and constructing individual path strings to test feels a bit wasteful. I wonder whether iterative calls to [`path.Split`](https://pkg.go.dev/path#Split) over a precomputed `path` string would make this logic easier to reason about.
</blockquote></details>

We can't use the `path` package because the path syntax may depend on the platform. We can't use `path/filepath` package either because the path syntax can be specific to the `vfs.FS` implementation (e.g. `MemFS`), and not match the platform syntax. We thus have to use the `vfs.FS` [helpers](https://github.com/cockroachdb/pebble/blob/51fca96df08fc8c95ba5b12208bad218c09a70dd/vfs/vfs.go#L162-L173) for manipulating paths. These and are somewhat underspecified in how they handle absolute/relative paths and root or '.' directories, so in my initial implementation I chose to not use them.

Changed the code to use the `fs.PathDir()` helper though, PTAL.
___
*[`pkg/kv/kvserver/logstore/sideload_disk.go` line 317 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/114616#gh-1409999778:-NkowAX0-hoJu8KUkDg1:b-mmd8m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/06a82d14da8ff900b392efc43e3fe893bbeb8b76/pkg/kv/kvserver/logstore/sideload_disk.go#L317)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Walking all the way up to the Pebble root directory and not assuming that "auxiliary" exists makes sense to me.

Or alternatively we could remove the distinction between the base directory path and the sub-directories we are trying to create and just create+sync whatever is necessary without making assumptions. This is analogous to what MkdirAll does.
</blockquote></details>

Done the latter.
___
*[`pkg/kv/kvserver/logstore/sideload_test.go` line 594 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/114616#-NkWi2gF-lFjAUdcb8rh:-NkoVCDA8tqyVJYzZWf5:b-8ubzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/72995ed527b32d819a672b79d73116dffa3124a3/pkg/kv/kvserver/logstore/sideload_test.go#L594)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why is this needed? Does `eng.Close()` try to sync before closing?
</blockquote></details>

The idea of this test is that we discard all the writes/syncs below, including the ones that `eng.Close()` might potentially do to gracefully terminate things. We are emulating a crash (see the comment below).

Correspondingly, calling `ss.Sync()` makes a difference to whether the entry persists or not (see the asserts below).


<!-- Sent from Reviewable.io -->

erikgrinaker commented on pkg/kv/kvserver/logstore/sideload_disk.go:
```diff
@@ -285,3 +302,48 @@ func (ss *DiskSideloadStorage) String() string {
+		}
+		parent = fs.PathDir(dir)
+		// NB: not checking against the separator, to be platform-agnostic.
+		if dir == "." || parent == dir { // reached the topmost dir or the root
```
~~Instead of special-casing `.` (and what about `..`?), we could simply call `filepath.Abs()` on the path first, which will resolve relative paths and normalize out any `.` and `..` components.~~

Nevermind, I see your comment above that we have to rely on Pebble helpers.

pavelkalinnikov replied:
I actually have no idea how these helpers behave with paths containing `..`, or in general with non-canonical paths. These are extremely brittle, and I think we should not use them. It would be cleaner to have a type-safe "canonical path" notion which is known to be relative to the data directory root which exists.

pavelkalinnikov replied:
I think the assumptions are:

- Pebble root directory exists.
- All paths within this directory are well-formed / canonical suffixes of the pebble root path.
- We never walk above the pebble root, so we're effectively always iterating nice canonical substrings.

All these things could be made explicit and constrained within a type-safe helper, so that the space of things we should worry about is reduced. Posted some ideas internally: https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1701873653835429?thread_ts=1700064004.728839&cid=CAC6K3SLU.

erikgrinaker commented on pkg/kv/kvserver/logstore/sideload_test.go:
```diff
@@ -574,3 +574,187 @@ func newOnDiskEngine(ctx context.Context, t *testing.T) (func(), storage.Engine)
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	for _, tc := range []struct {
```
Worth adding some test cases here with relative paths as well as intermediate `.` and `..` components.

pavelkalinnikov replied:
`MemFS` doesn't work well with those. For example, `a/b/c0/../c1` creates `a/b/c0`.

pavelkalinnikov replied:
Added some relative paths and `..` things. Also added a TODO to make these types of things safer.

erikgrinaker reviewed (approved):
Sorry for the hold-up, appreciate the patience.

erikgrinaker commented:
How do you feel about backporting this to 23.2, but leaving the older backports to bake for a while? That should give us a month of baking time for 23.2, which seems sufficient.

pavelkalinnikov commented:
blathers backport 23.2.0-rc

pavelkalinnikov commented:
blathers backport 23.1 22.2

pav-kv commented:
blathers backport 23.1
